src/plugin/vault/vault_test.go

```diff
@@ -225,51 +226,188 @@ func TestLogin(t *testing.T) {
 	}
 
-	time.Sleep(time.Duration(secret.Auth.LeaseDuration+1) * time.Second)
+	// Wait for TTL to expire and check that token is no longer valid
+	time.Sleep(time.Duration(secret.LeaseDuration+1) * time.Second)
+	_, err = c.AuthAPIClient.GetAdmins(c.Ctx(), &auth.GetAdminsRequest{})
+	if err == nil {
+		t.Fatalf("API call should fail, but token did not expire")
+	}
+}
+
+// TestLoginTTLParam tests that if TTL is set in the request, make sure the
+// returned token has that TTL
+func TestLoginTTLParam(t *testing.T) {
+	vaultClientConfig := vault.DefaultConfig()
+	vaultClientConfig.Address = vaultAddress
+	v, err := vault.NewClient(vaultClientConfig)
+	if err != nil {
+		t.Fatalf(err.Error())
+	}
+	v.SetToken("root")
+
+	err = configurePlugin(t, v, "")
+	if err != nil {
+		t.Fatalf(err.Error())
+	}
+
+	vl := v.Logical()
+	secret, err := vl.Write(
+		fmt.Sprintf("/%v/login/github:bogusgithubusername", pluginName),
+		map[string]interface{}{"ttl": "2s"},
+	)
+	if err != nil {
+		t.Fatalf(err.Error())
+	}
+	if secret.LeaseDuration != 2 {
+		t.Fatalf("Expected pachyderm token with TTL=2s, but was %ds", secret.LeaseDuration)
+	}
+	pachToken := secret.Data["user_token"].(string)
+	reportedPachdAddress := secret.Data["pachd_address"].(string)
+	c, err := client.NewFromAddress(reportedPachdAddress)
+	if err != nil {
+		t.Fatalf(err.Error())
+	}
+	c.SetAuthToken(pachToken)
+	// Make sure token is valid
+	_, err = c.AuthAPIClient.GetAdmins(c.Ctx(), &auth.GetAdminsRequest{})
+	if err != nil {
+		t.Fatalf(err.Error())
+	}
+	// Wait for TTL to expire and check that token is no longer valid
+	time.Sleep(time.Duration(secret.LeaseDuration+1) * time.Second)
 	_, err = c.AuthAPIClient.GetAdmins(c.Ctx(), &auth.GetAdminsRequest{})
 	if err == nil {
 		t.Fatalf("API call should fail, but token did not expire")
 	}
 }
 
+// If TTL is not in the request OR the config, make sure a sensible default is
+// used
+func TestLoginDefaultTTL(t *testing.T) {
+	c, _, secret := loginHelper(t, "")
+	if secret.LeaseDuration < 600 {
+		t.Fatalf("Expected Pachyderm token with duration at least 10m, but actual "+
+			"duration was %ds", secret.LeaseDuration)
+	}
+
+	// Make sure token is valid
+	_, err := c.AuthAPIClient.GetAdmins(c.Ctx(), &auth.GetAdminsRequest{})
+	if err != nil {
+		t.Fatalf(err.Error())
+	}
+}
+
+// TestLoginLongTTL tests that it's possible to get a Pachyderm token with a
+// long TTL
+func TestLoginLongTTL(t *testing.T) {
+	// Get Vault client
+	vaultClientConfig := vault.DefaultConfig()
+	vaultClientConfig.Address = vaultAddress
+	v, err := vault.NewClient(vaultClientConfig)
+	if err != nil {
+		t.Fatalf(err.Error())
+	}
+	v.SetToken("root")
+	vl := v.Logical()
+
+	// Get Pachyderm token with long TTL
+	secret, err := vl.Write(
+		fmt.Sprintf("/%v/login/github:bogusgithubusername", pluginName),
+		map[string]interface{}{"ttl": "768h"},
+	)
+	if err != nil {
+		t.Fatalf(err.Error())
+	}
+	if secret.LeaseDuration < int((700 * time.Hour).Seconds()) {
```

Comment:
Makes sense we want to measure a bound, but we don't expect 68 hours to have passed during this test ... so I think we can tighten up this value a bit.

Reply:
I actually think 700h is the right value here. Pachyderm used to reject leases longer than 2 weeks, but it should accept leases up to 30d (720h. Go time durations do not allow days) which is what this checks. I'll rewrite 700 to 30*24-1